### DISCUSSION TITLE
Deserialize/Unmarshal JSON String to different structs

### DISCUSSION THREAD

**Discussion by rohithdv** [fybrik/fybrik#1029] [Developers]
I had a clarification on deserialization using the data catalog taxonomy. The following example will provide the details. 

There are 2 structs which are generated from data catalog taxonomy:

Connection1: 
// Connection struct for Connection
type Connection struct {
    Name                 string `json:"name"`
    AdditionalProperties map[string]interface{} 
} 
Connection1 is generated when we just have base.yaml. There is no example taxonomy. This is probably what the manager needs to use. It does not need to know about s3 or db2 or any other connection type.

Connection2 is generated when we combine base.yaml with example taxonomy (for example use s3).

Connection2: 
type Connection2 struct {
    S3   *S3    `json:"S3,omitempty"`
    Name string `json:"name"`
}

type S3 struct {
    Bucket    string  `json:"bucket"`
    Endpoint  string  `json:"endpoint"`
    ObjectKey string  `json:"objectKey"`
    Region    *string `json:"region,omitempty"`
}

Now suppose we have a connector (which knows about Connection2) which serializes a Connection json string as below. 
JSON STRING: 
{
        "name": "s3",
        "s3": {
            "Endpoint":  "s3.eu-gb.cloud-object-storage.appdomain.cloud",
            "Bucket":    "fybrik-test-bucket",
            "ObjectKey": "test.csv"
        }
 }

Q: How can we deserialize the  above json string to Connection1 struct when the manager receives it ? 
@roee88 @roytman @ronenkat @shlomitk1 @ritwikc

**Comment by shlomitk1** [comment]
Manager does not need to deserialize back to Connection1.
The problem I got while attempting to use the Connection structure in the same way Action is used was that somehow an attribute "AdditionalProperties" was added to the structure, sometimes on top of another AdditionalProperties.

**Reply by rohithdv** [reply]
Ok I see..Just for my understanding how is Connection1 used by the manager ? I was thinking that is the way the manager stores the connection information. Thanks.

**Reply by roee88** [reply]
> Q: How can we deserialize the above json string to Connection1 struct when the manager receives it ?

Using json.Unmarshal should just work. What problem do you see?


> Manager does not need to deserialize back to Connection1.

Do you mean because it uses it as raw json and you extract the name manually? 

> The problem I got while attempting to use the Connection structure in the same way Action is used was that somehow an attribute "AdditionalProperties" was added to the structure, sometimes on top of another AdditionalProperties.

I don't understand this comment given how Marshal/Unmarshal handles additional properties in https://github.com/fybrik/fybrik/blob/ac8ea647b849bb028cbea636e86422507197c734/pkg/taxonomy/model/datacatalog/base/model_connection.go#L80. 

@rohithdv it would be good if you can describe the exact problem that you're having.

**Reply by rohithdv** [reply]
Sorry, I maybe misunderstanding some technical aspects here. For explaining purposes, I have attached the following 2 files: deserialization.go.txt (it is a go file. Please remove the .txt extension. github was not allowing me to attach go files. so I have added a .txt extension) and output-deserialization.txt (which is the output I get if I run the previous go file). 
[deserialization.go.txt](https://github.com/fybrik/fybrik/files/7717721/deserialization.go.txt)
[output-deserialization.txt](https://github.com/fybrik/fybrik/files/7717720/output-deserialization.txt)

My clarification is how to deserialize  to `Connection` struct given in the above example. I am able to deserialize to `Connection2` and `Connection3` structs but not to `Connection` struct. Please let me know your views and thoughts regarding this. Thanks a lot.  @roee88

**Reply by roee88** [reply]
I'll take a look. @rohithdv when deserialization.go.txt includes the UnmarshalJSON function form 
https://github.com/fybrik/fybrik/blob/ac8ea647b849bb028cbea636e86422507197c734/pkg/taxonomy/model/datacatalog/base/model_connection.go#L80, does it still give that empty output?

**Reply by rohithdv** [reply]
When I use UnmarshalJSON I get the following output @roee88 
```

Output of deserialization to Connection struct using  UnmarshalJSON:{
        "name": "Kafka",
        "AdditionalProperties": {
                "kafka": {
                        "bootstrap_servers": "http://kafka-servers",
                        "key_deserializer": "io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer",
                        "name": "Kafka",
                        "sasl_mechanism": "SCRAM-SHA-512",
                        "schema_registry": "kafka-registry",
                        "security_protocol": "SASL_SSL",
                        "ssl_truststore": "xyz123",
                        "ssl_truststore_password": "passwd",
                        "topic_name": "topic",
                        "value_deserializer": "io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer"
                }
        }
}
```

**Reply by rohithdv** [reply]
I think I maybe wrong in understanding the json.unmarshal function..With the following changes we dont get null in any scenario! I think the unmarshal function will call the unmashalJSON internally somehow (if it is defined)..so using just unmarshal function would still work as @roee88 mentioned!
 
[deserialization.go.txt](https://github.com/fybrik/fybrik/files/7720327/deserialization.go.txt)
[output-deserialization.txt](https://github.com/fybrik/fybrik/files/7720333/output-deserialization.txt)

**Comment by roytman** [comment]
Hi @rohithdv , I probably the issue. If you use the UnmarshalJSON from fybrik/pkg/taxonomy/model/datacatalog/base/model_connection.go, you get something like:
```
2021/12/15 11:52:05 Output of deserialization to Connection struct :{
	"name": "Kafka",
	"AdditionalProperties": {
		"kafka": {
			"bootstrap_servers": "http://kafka-servers",
			"key_deserializer": "io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer",
			"name": "Kafka",
			"sasl_mechanism": "SCRAM-SHA-512",
			"schema_registry": "kafka-registry",
			"security_protocol": "SASL_SSL",
			"ssl_truststore": "xyz123",
			"ssl_truststore_password": "passwd",
			"topic_name": "topic",
			"value_deserializer": "io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer"
		}
	}
}
```
in order to eliminate the `kafka` from AdditionalProperties, you can change the desirialization to something like :
 ````
 func (o *Connection) UnmarshalJSON(bytes []byte) (err error) {
	varConnection := make(map[string]interface{})

	if err = json.Unmarshal(bytes, &varConnection); err != nil {
		// do some logging; return error
	}
	o.Name = varConnection["name"].(string) // we can check that the value exist and is String
	delete(varConnection, "name")
	if len(varConnection) != 1 {
		// do some logging; return error
	} else {
		o.AdditionalProperties = make(map[string]interface{})
		for key, value := range varConnection {
			var ok bool
			o.AdditionalProperties, ok = value.(map[string]interface{})
			if !ok {
				// do some logging; return error
			}
			// optionally we can add
			o.AdditionalProperties["type"] = key
		}
	}
	return err
}
```

(As a side performance effect it doesn't have 2 desirializations)
Its output is 
```
2021/12/15 12:14:58 Output of deserialization to Connection struct :{
	"name": "Kafka",
	"AdditionalProperties": {
		"bootstrap_servers": "http://kafka-servers",
		"key_deserializer": "io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer",
		"name": "Kafka",
		"sasl_mechanism": "SCRAM-SHA-512",
		"schema_registry": "kafka-registry",
		"security_protocol": "SASL_SSL",
		"ssl_truststore": "xyz123",
		"ssl_truststore_password": "passwd",
		"topic_name": "topic",
		"type": "kafka",
		"value_deserializer": "io.confluent.kafka.serializers.json.KafkaJsonSchemaDeserializer"
	}
}
```

**Reply by rohithdv** [reply]
Thanks @roytman . I think this approach will work. My point was that a simple unmarshal function may not work. It is better to use the unmarshalJSON ( I had overlooked this function) or a customized version of it to get the deserialization done. Thanks for the detailed clarification.

**Reply by roee88** [reply]
But why would you want to eliminate the "Kafka" key?

**Reply by rohithdv** [reply]
Actually regarding the "Kafka" key point which @roee88  asked above, I am facing a similar issue in another context. It would be great to get your views on this. 

We have auto-generated [`ResourceColumns` struct](https://github.com/rohithdv/fybrik/blob/4d03458f2412d59ad55d4f4607a50945326cfd48/pkg/taxonomy/model/datacatalog/base/model_resource_columns.go#L20) while defining models for Data Catalog based on the `columns` field under `Resource` field in  [data_catalog.json](https://github.com/rohithdv/fybrik/blob/4d03458f2412d59ad55d4f4607a50945326cfd48/charts/fybrik/files/taxonomy/data_catalog.json#L42). Now if you see the `Tags` field in the [`ResourceColumns` struct](https://github.com/rohithdv/fybrik/blob/4d03458f2412d59ad55d4f4607a50945326cfd48/pkg/taxonomy/model/datacatalog/base/model_resource_columns.go#L20)  it is a pointer to `map[string]interface{}` (similar to `AdditionalProperties` field in the above conversations). The problem is while using this field, we need to define an `extra` key before providing the actual tags due to the `*map[string]interface{}` type as shown in the example below. Is there any way to avoid the extra key field (in the below example I have used the string "tags1" but it could be any other string) ? Thanks.  @roee88 @roytman 
```
	"resource": {
		"columns": [
			{
				"name": "nameOrig",
				"tags": {
					"tags1": [
						"PII"
					]
				}
			},
			{
				"name": "newbalanceOrig",
				"tags": {
					"tags1": [
						"sensitive"
					]
				}
			},
			{
				"name": "oldbalanceOrg",
				"tags": {
					"tags1": [
						"sensitive"
					]
				}
			}
		],
		"geography": "theshire",
		"name": "fybrik-notebook-sample/data-csv",
		"owner": "",
		"tags": {
			"tags1": [
				"finance"
			]
		}
	}
```
